qldpc/abstract.py

```diff
@@ -525,3 +598,110 @@ def lift(member: int) -> IntegerArray:
+    def get_generator_mats(
+        cls, dimension: int, field: int | None = None
+    ) -> tuple[galois.FieldArray, galois.FieldArray]:
+        """Generator matrices for the special linear group SL(dimension, field).
+
+        This construction is based on https://arxiv.org/abs/2201.09155.
+        """
+        base_field = galois.GF(field or DEFAULT_FIELD_ORDER)
+        A = base_field.Identity(dimension)
+        W = -np.roll(base_field.Identity(dimension), 1, axis=0)
+        W[0, -1] = 1
+        if base_field.order > 3:
+            A[0, 0] = base_field.primitive_element
+            A[1, 1] = base_field.primitive_element**-1
+            W[0, 0] = -1 * base_field(1)
+        else:
+            A[0, 1] = 1
+        return A, W
+
+
+def ProjectiveSpecialLinearGroup(Group):
+    """Projective variant of the special linear group."""
+
+    def __init__(self, dimension: int, field: int | None = None) -> None:
+        base_field = galois.GF(field or DEFAULT_FIELD_ORDER)
+        if base_field.order == 2:
+            super().__init__(SpecialLinearGroup(dimension, 2))
+        elif dimension == 2:
+            generators = self.get_expanding_generator_mats(dimension, field)
+            space = _construct_projspace(dimension, field)
+            group = Group.from_permutation_mats(generators, space, field)
+            super().__init__(group)
+        else:
+            raise ValueError(
+                "Projective special linear groups with both dimension and field greater than 2 are"
+                " not yet supported."
+            )
+
+    @classmethod
+    def get_expanding_generator_mats(
+        cls, field: int | None = None
+    ) -> tuple[galois.FieldArray, galois.FieldArray]:
+        """Expanding generator matrices for PSL(2, field), from https://arxiv.org/abs/1807.03879"""
+        base_field = galois.GF(field or DEFAULT_FIELD_ORDER)
+        minus_one = -base_field(1)
+        A = base_field([[1, 1], [0, 1]])
+        B = base_field([[1, minus_one], [0, 1]])
+        C = base_field([[1, 0], [1, 1]])
+        D = base_field([[1, 0], [minus_one, 1]])
+        return A, B, C, D
+
+
+def _construct_linspace(dimension: int, field: int | None = None) -> list[bytes]:
+    """Helper function to generate a list of vectors over finite field."""
+    base_field = galois.GF(field or DEFAULT_FIELD_ORDER)
+    vectors = itertools.product(base_field.elements, repeat=dimension)
+    next(vectors)  # skip the all-0 element
+    return [base_field(vec).tobytes() for vec in vectors]
+
+
+def _construct_projspace(dimension: int, field: int) -> list[bytes]:
+    """Helper function to create the vectors in the projective space.
+    The difference from usual vectors is that scalar multiples are identified.
+    """
+    base_field = galois.GF(field)
+    return [
+        base_field(vec).tobytes()
+        for vec in itertools.product(base_field.elements, repeat=dimension)
+        if vec[(base_field(vec) != 0).argmax()] == 1
+    ]
```

Comment:
I cleaned these up a bit, please make sure that they still look correct to you.

Reply:
They look correct (and clean :)